### PR TITLE
[mrp] MO & WO & Pickings - Raise constraint if wrong SN encoded

### PR DESCRIPTION
**Pull request by hiralvbhavsar**
Description of the issue/feature this PR addresses:

Task: https://www.odoo.com/web?#id=36389&view_type=form&model=project.task&action=333&active_id=131&menu_id=4720
Pad: https://pad.odoo.com/p/r.95462cd4392604d48a6ebb309d9a8ecb

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/stock/models/stock_move_line.py [hunk 1]
```diff
@@ -98,8 +98,52 @@ def onchange_product_id(self):
...
+
+    def _get_move_lines(self):
+        self.ensure_one()
+        lines = self.env['stock.move.line']
+        if self.move_id.picking_id:
+            lines |= self.move_id.picking_id.move_line_ids.filtered(lambda ml: ml.product_id == self.product_id and (ml.lot_id or ml.lot_name))
+        return lines
+
+    def _check_lot_exists(self):
+        seen = set()
```
Comment (sle-odoo) [inline addons/stock/models/stock_move_line.py]: """ This method is used in `_check_unique_serial_number` and in `onchange_serial_number` to check that a same serial number is not used twice amongst the recordset passed.

:return: an error message directed to the user if needed else False
"""

File: addons/stock/models/stock_move_line.py [hunk 2]
```diff
@@ -98,8 +98,52 @@ def onchange_product_id(self):
...
+                    return _('You have already assigned this serial number to this product. Please correct the serial numbers encoded.')
+
+    def _get_move_lines(self):
+        self.ensure_one()
+        lines = self.env['stock.move.line']
+        if self.move_id.picking_id:
+            lines |= self.move_id.picking_id.move_line_ids.filtered(lambda ml: ml.product_id == self.product_id and (ml.lot_id or ml.lot_name))
+        return lines
+
+    def _check_lot_exists(self):
```
Comment (sle-odoo) [inline addons/stock/models/stock_move_line.py]: i would rename this method into something more sensible, like `_check_for_duplicated_serial_numbers` or `_check_serial_numbers_usage`

File: addons/stock/models/stock_move_line.py [hunk 3]
```diff
@@ -98,8 +98,52 @@ def onchange_product_id(self):
...
+        seen = set()
+        for move_line in self:
+            if move_line.lot_id:
```
Comment (sle-odoo) [inline addons/stock/models/stock_move_line.py]: something like this (untested)
```
recorded_serials_counter = Counter(self.mapped('lot_id.name'))
for lot_id, occurrences in recorded_serials_counter.items():
    if occurrences > 1:
        raise...
```
would be clearer don't you think?

File: addons/stock/models/stock_move_line.py [hunk 4]
```diff
@@ -98,8 +98,52 @@ def onchange_product_id(self):
...
+            if float_compare(ml.qty_done, 1.0, precision_rounding=ml.move_id.product_id.uom_id.rounding) == 1:
+                raise UserError(_(
+                    'You can only process 1.0 %s for products with unique serial number.') % ml.product_id.uom_id.name)
+            if ml.lot_name:
+                already_exist = self.env['stock.production.lot'].search(
+                    [('name', '=', ml.lot_name), ('product_id', '=', ml.product_id.id)])
+                if already_exist or float_compare(ml.qty_done, 1.0, precision_rounding=ml.move_id.product_id.uom_id.rounding) == 1:
+                    return _('You have already assigned this serial number to this product. Please correct the serial numbers encoded.')
+
+    def _get_move_lines(self):
```
Comment (sle-odoo) [inline addons/stock/models/stock_move_line.py]: there's already a method named `_get_move_lines`on stock.move that return the move lines to consider for _quantity_done_compute, i suggest to change this method name for clarity sake

File: addons/stock/models/stock_move_line.py [hunk 5]
```diff
@@ -98,8 +98,52 @@ def onchange_product_id(self):
 
     @api.onchange('lot_name', 'lot_id')
     def onchange_serial_number(self):
+        res = {}
         if self.product_id.tracking == 'serial':
             self.qty_done = 1
+            move_lines_to_check = self._get_move_lines() - self._origin
```
Comment (sle-odoo) [inline addons/stock/models/stock_move_line.py]: add a comment to explain why you have to remove yourself

File: addons/stock/models/stock_move_line.py (hunk 2 above)
Comment (sle-odoo) [reply addons/stock/models/stock_move_line.py]: or something not beginning by _check as one could think it's a constraint

File: addons/stock/models/stock_move_line.py [hunk 6]
```diff
@@ -98,8 +98,52 @@ def onchange_product_id(self):
...
+            message = move_lines_to_check._check_lot_exists()
+            if message:
+                res['warning'] = {'title': _('Warning'), 'message': message}
+        return res
+
+    @api.constrains('lot_id', 'lot_name', 'qty_done')
+    def _check_unique_serial_number(self):
+        move_lines_serial = self.filtered(lambda ml: ml.move_id.product_id.tracking == 'serial' and (ml.lot_id or ml.lot_name))
+        for ml in move_lines_serial:
+            move_lines_to_check = ml._get_move_lines() - ml
```
Comment (pimodoo) [inline addons/stock/models/stock_move_line.py]: maybe should not remove it

File: addons/stock/models/stock_move_line.py [hunk 7]
```diff
@@ -98,8 +98,52 @@ def onchange_product_id(self):
...
+            message = move_lines_to_check._check_lot_exists()
+            if message:
+                raise ValidationError(message)
+            if float_compare(ml.qty_done, 1.0, precision_rounding=ml.move_id.product_id.uom_id.rounding) == 1:
+                raise UserError(_(
+                    'You can only process 1.0 %s for products with unique serial number.') % ml.product_id.uom_id.name)
+            if ml.lot_name:
+                already_exist = self.env['stock.production.lot'].search(
+                    [('name', '=', ml.lot_name), ('product_id', '=', ml.product_id.id)])
+                if already_exist or float_compare(ml.qty_done, 1.0, precision_rounding=ml.move_id.product_id.uom_id.rounding) == 1:
```
Comment (sle-odoo) [inline addons/stock/models/stock_move_line.py]: second part of the if is already taken into account at line 118

File: addons/mrp/wizard/mrp_product_produce.py [hunk 8]
```diff
@@ -175,5 +176,25 @@ class MrpProductProduceLine(models.TransientModel):
...
+            lots = set([])
+            # Check duplication of the serial number.
+            for move in self.product_produce_id.produce_line_ids.filtered(lambda l: l.lot_id) - self:
+                if move.lot_id in lots:
+                    res['warning'] = {'title': _('Warning'), 'message': _('You cannot consume the same serial number twice. Please correct the serial numbers encoded.')}
+                    return res
+                lots.add(move.lot_id)
+        return res
+
+    @api.one
```
Comment (sle-odoo) [inline addons/mrp/wizard/mrp_product_produce.py]: not sure

File: addons/stock/models/stock_move_line.py [hunk 9]
```diff
@@ -98,8 +98,52 @@ def onchange_product_id(self):
...
             self.qty_done = 1
+            move_lines_to_check = self._get_move_lines() - self._origin
+            message = move_lines_to_check._check_lot_exists()
+            if message:
+                res['warning'] = {'title': _('Warning'), 'message': message}
+        return res
+
+    @api.constrains('lot_id', 'lot_name', 'qty_done')
+    def _check_unique_serial_number(self):
+        move_lines_serial = self.filtered(lambda ml: ml.move_id.product_id.tracking == 'serial' and (ml.lot_id or ml.lot_name))
```
Comment (sle-odoo) [inline addons/stock/models/stock_move_line.py]: no need to store it in a variable

File: addons/mrp/wizard/mrp_product_produce.py [hunk 10]
```diff
@@ -175,5 +176,25 @@ class MrpProductProduceLine(models.TransientModel):
 
     @api.onchange('lot_id')
     def _onchange_lot_id(self):
+        res = {}
         if self.product_id.tracking == 'serial':
             self.qty_done = 1
+            lots = set([])
+            # Check duplication of the serial number.
+            for move in self.product_produce_id.produce_line_ids.filtered(lambda l: l.lot_id) - self:
```
Comment (sle-odoo) [inline addons/mrp/wizard/mrp_product_produce.py]: i would prefer a collections.Counter here too. but if you want to keep it this way it's ok too.

**Comment by pimodoo** [comment]
It has been merged in saas-18, thanks for your job !

**Pull request closed without merging**